pkg/controller/supportbundlecollection/controller.go

```diff
@@ -0,0 +1,1016 @@
+}
+
+// addInternalSupportBundleCollection adds internalBundle into supportBundleCollectionStore, and creates a
+// supportBundleCollectionAppliedTo resource to maintain the SupportBundleCollection's required Nodes or ExternalNodes.
+func (c *Controller) addInternalSupportBundleCollection(
+	bundleCollection *v1alpha1.SupportBundleCollection,
+	nodeSpan sets.String,
+	authentication controlplane.BundleServerAuthConfiguration,
+	expiredAt metav1.Time) error {
+	var nodeNames, enNames []string
+	var nodeSelector, enSelector labels.Selector
+	var err error
+	var processNodes bool
+	if bundleCollection.Spec.Nodes == nil {
+		processNodes = false
+		nodeSelector = labels.Nothing()
+		nodeNames = []string{}
+	} else {
+		processNodes = true
+		nodeNames = bundleCollection.Spec.Nodes.NodeNames
+		if bundleCollection.Spec.Nodes.NodeSelector == nil {
+			nodeSelector = labels.Nothing()
+		} else {
+			nodeSelector, err = metav1.LabelSelectorAsSelector(bundleCollection.Spec.Nodes.NodeSelector)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	var enNamespace string
+	if bundleCollection.Spec.ExternalNodes == nil {
+		enNamespace = ""
+		enSelector = labels.Nothing()
+		enNames = []string{}
+	} else {
+		enNamespace = bundleCollection.Spec.ExternalNodes.Namespace
+		enNames = bundleCollection.Spec.ExternalNodes.NodeNames
+		if bundleCollection.Spec.ExternalNodes.NodeSelector == nil {
+			enSelector = labels.Nothing()
+		} else {
+			enSelector, err = metav1.LabelSelectorAsSelector(bundleCollection.Spec.ExternalNodes.NodeSelector)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	appliedTo := &supportBundleCollectionAppliedTo{
+		name:         bundleCollection.Name,
+		nodeNames:    nodeNames,
+		nodeSelector: nodeSelector,
+		processNodes: processNodes,
+		enNamespace:  enNamespace,
+		enNames:      enNames,
+		enSelector:   enSelector,
+	}
+	c.supportBundleCollectionAppliedToStore.Add(appliedTo)
+	// Create internal SupportBundleCollection resource.
+	internalBundleCollection := &types.SupportBundleCollection{
+		SpanMeta: types.SpanMeta{
+			NodeNames: nodeSpan,
+		},
+		Name:           bundleCollection.Name,
+		UID:            bundleCollection.UID,
+		SinceTime:      bundleCollection.Spec.SinceTime,
+		FileServer:     bundleCollection.Spec.FileServer,
+		ExpiredAt:      expiredAt,
+		Authentication: authentication,
+	}
+	_ = c.supportBundleCollectionStore.Create(internalBundleCollection)
+	return nil
+}
+
+func (c *Controller) addNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	affectedBundleCollections := c.filterSupportBundleCollectionsFromNode(node, false)
+	for bundle := range affectedBundleCollections {
+		c.queue.Add(bundle)
+		klog.InfoS("Enqueued SupportBundleCollection by Node ADD event", "name", bundle)
+	}
+	klog.V(2).InfoS("Processed Node Add event", "name", node.Name)
+}
+
+func (c *Controller) updateNode(oldObj interface{}, newObj interface{}) {
```

Comment:
Do we need to handle Node updates? Could we ignore Node updates after CRD is processed for simplicity? Anyway, we have no way to guarantee all Node updates are covered as update events are async.

Reply:
I removed updateNode and updateExternalNode.